In [164]:
!pip install haversine

  Running setup.py bdist_wheel for haversine ... done
  Stored in directory: /Users/rajchakrabarty/Library/Caches/pip/wheels/be/a3/4e/2e073dc725bf9b430e2d42931c55e5be2be4cdc6e325e17677
Successfully built haversine
You are using pip version 9.0.1, however version 9.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


In [1]:
import keys
import pandas as pd
import requests
from pprint import pprint
import json
import numpy as np

In [3]:
screens = pd.read_csv('screen geocoded.csv')
screens.drop('Unnamed: 0', axis=1, inplace=True)
screens.shape

(628, 12)

In [4]:
screens.head()

Screen Design: Screen Design  \
0                             Azure Park (JPM & ARP)   
1  Banyan Street Capital - ATL - 191 Peachtree Tower   
2                    athenahealth (WeDriveU) Atlanta   
3                           Buckhead at Lenox Square   
4           Seven Oaks - ATL - 4004 Perimeter Summit   

                       Opportunity: Opportunity Name  \
0                    JPM & ARP-Azure on the Park ATL   
1  Banyan Street Capital - ATL - 191 Peachtree Tower   
2    WeDriveU-AthenaHealth - Arsenal Campus- Revisit   
3   Parmenter - ATL - Buckhead Tower at Lenox Square   
4           Seven Oaks - ATL - 4004 Perimeter Summit   

               Screen Placement Address Screen Addresss (City)  \
0                  1020 Piedmont Ave NE                Atlanta   
1                  191 Peachtree St. NE                Atlanta   
2  675 Ponce de Leon Ave NE, Suite 9000                Atlanta   
3                3399 Peachtree Road NE                Atlanta   
4            4004 Perimeter Summit Pkwy                Atlanta   

  Screen Address (State) Screen Address (Zip)  Install Stage Region  \
0                     GA                30309  Customer Live    ATL   
1                     GA                30303  Customer Live    ATL   
2                     GA                30308  Customer Live    ATL   
3                     GA                30326  Customer Live    ATL   
4                     GA                30319  Customer Live    ATL   

     Screen Type        lat        lng  \
0  Public Screen  33.782524 -84.380599   
1  Public Screen  33.749514 -84.394477   
2  Public Screen  33.772828 -84.365784   
3  Public Screen  33.846098 -84.359549   
4  Public Screen  33.915213 -84.340141   

                                             Address  
0              1020 Piedmont Ave NE Atlanta GA 30309  
1              191 Peachtree St. NE Atlanta GA 30303  
2  675 Ponce de Leon Ave NE, Suite 9000 Atlanta G...  
3            3399 Peachtree Road NE Atlanta GA 30326  
4        4004 Perimeter Summit Pkwy Atlanta GA 30319

In [5]:
#look at number that are missing lats and longs
print (screens[screens.lat.isnull()==True].shape)
print (screens[screens.lat.isnull()==False].shape)
print ()

(51, 12)
(577, 12)



In [ ]:
########################################

In [7]:
#add close date from model-data file
model_data = pd.read_csv('model_data.csv')

In [8]:
model_data.head()

id                                            name close_date  \
0  0061a00000GcQdL                       525 Water St SW- CMC Mgmt     3/8/18   
1  0061a00000Mi8dC                    Bozzuto - NJ - Park + Garden    6/30/18   
2  0061a00000M4PAU                           Greystar- The Cameron   11/17/17   
3  0061a00000MgelJ                          JLL- 2035 NW Front Ave    1/31/18   
4  0061a00000KOF8z  International Monetary Fund- Spring Conference    1/31/18   

           stage                                            address  \
0  Contract Sent  525 Water St SW Washington United States Distr...   
1   Invoice Sent  1450 Garden St Hoboken United States New Jerse...   
2    Deal Signed           8710 Cameron St, Silver Spring, MD 20910   
3  Proposal Sent          1120 NW Couch St #500, Portland, OR 97209   
4     Closed Won     1900 Pennsylvania Ave NW, Washington, DC 20431   

         lat         lng                                        ts_response  \
0  38.876059  -77.020240  {'data': {'mobilityScore': {'latitude': 38.876...   
1  40.754448  -74.028046  {'data': {'mobilityScore': {'latitude': 40.754...   
2  38.999289  -77.028885  {'data': {'mobilityScore': {'latitude': 38.999...   
3  45.537557 -122.688094  {'data': {'mobilityScore': {'latitude': 45.537...   
4  38.898940  -77.044101  {'data': {'mobilityScore': {'latitude': 38.898...   

   mobility_score   carshare  bikeshare  ridehailing  masstransit  \
0            86.0  12.190237  21.710327    21.000000    25.347332   
1            85.0   3.435388  25.069638    15.000000    33.883246   
2            91.0   2.921386   8.266361    12.115385    70.074879   
3           100.0   0.000000   0.000000    21.000000     1.468259   
4           100.0  14.075039  40.000111    21.000000    87.423132   

                                     lat_lng    closest_ts  within_one_tenth  \
0  (38.876059000000012, -77.020239899999979)  7.073231e-13               1.0   
1  (40.754448199999999, -74.028046000000003)  1.927253e-01               0.0   
2          (38.9992886, -77.028885400000021)  2.301765e-01               0.0   
3         (45.537556700000003, -122.6880937)  6.848705e-01               0.0   
4  (38.898939599999999, -77.044101499999996)  2.050043e-01               0.0   

   within_one_half  within_one  within_five  
0              6.0        11.0         72.0  
1              1.0         3.0         31.0  
2              4.0         6.0         11.0  
3              0.0         1.0          6.0  
4              4.0        11.0         92.0

In [9]:
close_dates = model_data[['name'
                          ,'close_date'
                          #,'address'
                         ]]

In [10]:
close_dates.shape

(1683, 2)

In [11]:
screens = screens.merge(close_dates, how='left'
                        ,right_on='name'
                        ,left_on='Opportunity: Opportunity Name'
                        #,indicator=True
                       )

In [12]:
#why did the number of records go from 628 to 649? are there dups?
print(screens.shape)
print(screens[screens.close_date.isnull()==False].shape)
print(screens[screens.close_date.isnull()==True].shape)

(649, 14)
(407, 14)
(242, 14)


In [14]:
screens.drop(['name'], axis=1, inplace=True)

In [13]:
screens[screens.close_date.isnull()==True].head(100)

Screen Design: Screen Design  \
2                      athenahealth (WeDriveU) Atlanta   
5     Engrain - Post Alexander - ATL - 600 Phipps Blvd   
9                    Local on 14th- Matrix Residential   
11            Seven Oaks - ATL - 3003 Perimeter Summit   
12            Seven Oaks - ATL - 1001 Perimeter Summit   
16            Seven Oaks - ATL - 2002 Perimeter Summit   
25                   Halstead Malden Square Apartments   
26                 Watermark Seaport - 85 Seaport Blvd   
28                           MIT School of Engineering   
29         Cresset Group - 65 Grove Street - Watertown   
32                     Hanover Company - BOS - Alewife   
38                              Harvard CS50 classroom   
45                 Equity Office - BOS - 100 Summer St   
46          Deco - Quincy Bozzuto - Quincy MA - 2 Penn   
47   Hobbs Office Park (128 Business Council) - Wal...   
50    Avalon-AVB Exeter - BOS - 77 Exeter St - TS Lite   
54   Avalon-AVA Somerville - Somerville MA - 445 Ar...   
57   Engrain-Fuse Cambridge - Cambridge MA - 165 Ca...   
58                                         CS50 at HSA   
59   Linea Cambridge - Cambridge MA - 1 Whittemore ...   
61                              225 Centre St- Peabody   
64                         Residences @ Malden Station   
65                    99 Tremont - Mainsail Management   
68                              Transwestern - 99 High   
69                                      Synlogic, Inc.   
71                      Harvard CS50 Pforzheimer House   
76    Biomed Realty - 200 Sidney Street - Cambridge MA   
78          Avalon Quincy - 151 Quarry Street - Boston   
84                          Harvard CS50 - Cabot House   
85                        Harvard CS50 - Currier House   
..                                                 ...   
196                            MakeOffices - Arlington   
197                             MakeOffices - Bethesda   
198                                   2201 Wilson Blvd   
201                             Marriott HQ - Bethesda   
205                    Bozzuto - Arlington - The Rixey   
208  DC - Takoma Spring Place - Keystar - 7051 Spri...   
213  Discovery HQ Silver Spring Screen P2 Exit on R...   
214                               1310 N Courthouse Rd   
221                      MakeOffices-DC- 1015 15 St NW   
227               NKGF - 1776 - Lobby 1133 15th street   
230       Carlyle Overlook [2318 Mill Rd - Alexandria]   
232  DCRA (DDOT) - DC - Waterfront Office - 1100 4t...   
235                            The Exchange at Wheaton   
238                       Wharf Parcel 2 - The Channel   
239  Wharf Parcel 3A Office- 800 Maine (Make Office...   
240                          Wharf Parcel 4A - INCANTO   
242                                     PlaceScoop ULI   
249                                         The Citron   
266                     City of Denver - Webb Building   
268                              Four Winds HQ- Denver   
273                   GM -East Lobby VEC (Tower Lobby)   
274  GM - Cadillac Bldg - Warren MI - GM Tech Cente...   
275                               Amazon Dublin Office   
276                          Energy Excelerator Office   
281  Avalon-AVA Burbank - Burbank CA - 401 N Pass A...   
282                     City of Santa Monica City Hall   
283                          TouchSource - UCLA Health   
286                  8th & Grand Apartments - Greystar   
297                                   7770 Norfolk/JBG   
298  JBG - The Alaire - Rockville MD - 1101 Higgins Pl   

                         Opportunity: Opportunity Name  \
2      WeDriveU-AthenaHealth - Arsenal Campus- Revisit   
5                                 Post Alexander - ATL   
9             Matrix Residential - ATL - Local on 14th   
11            Seven Oaks - ATL - 3003 Perimeter Summit   
12            Seven Oaks - ATL - 1001 Perimeter Summit   
16            Seven Oaks - ATL - 2002 Perimeter Summit   
25   Priderock Capital 

unmatching screen addresses:
WeDriveU-AthenaHealth - Arsenal Campus- Revisit

Post Alexander - ATL

Matrix Residential - ATL - Local on 14th

Seven Oaks - ATL - 3003 Perimeter Summit



In [135]:
model_data[model_data.lat==33.772828]

Empty DataFrame
Columns: [id, name, close_date, stage, address, lat, lng, ts_response, mobility_score, carshare, bikeshare, ridehailing, masstransit]
Index: []

It looks like this screen didn't exist in close dates

In [15]:
model_data[model_data.stage=='Closed Won'].shape

(295, 19)

In [156]:
screens.drop_duplicates(['lat'], keep='first', inplace=True)

In [154]:
screens.to_csv('screens_and_close_dates.csv', index=False)

In [ ]:
################################################

In [173]:
from haversine import haversine

In [256]:
def find_closest(point, list_of_points, debug=False):
    distances = []
    for i in list_of_points:
        #if it is a different point, calculate the distance
        if point != i:
            dist = haversine(point, i)
            if debug:
                print (point, i, dist)
            distances.append(dist)
    #return the smallest distance 
    #do we want to set a max here?
    return (distances[np.argmin(distances)])
    

In [258]:
point = (38.876059, -77.020240)
list_of_points = [(40.754448,-74.028046), (38.999289,-77.028885), (38.876059, -77.020240)]
find_closest(point, list_of_points,True)

(38.876059, -77.02024) (40.754448, -74.028046) 330.0218885526628
(38.876059, -77.02024) (38.999289, -77.028885) 13.722935990827294


13.722935990827294

In [215]:
point = (38.876059, -77.020240)
list_of_points = [(40.754448,-74.028046), (38.999289,-77.028885)]
find_closest(point, list_of_points)

(38.876059, -77.02024) (40.754448, -74.028046) 330.0218885526628
(38.876059, -77.02024) (38.999289, -77.028885) 13.722935990827294


13.722935990827294

In [327]:
def find_within_distance(point, list_of_points, radius=.5):
    count = 0
    for i in list_of_points:
        dist = haversine(point, i)
        #if point == i:
        #    print('found self', i)
        if point != i:
            if dist <= radius:
                count +=1
    return (count)

In [251]:
point = (38.876059, -77.020240)
list_of_points = [(40.754448,-74.028046), (38.999289,-77.028885)]
find_within_distance(point, list_of_points,14)

1

In [210]:
model_data.shape

(1683, 13)

In [214]:
print(screens.shape)
print(screens[screens.close_date.isnull()==False].shape)
print(screens[screens.close_date.isnull()==True].shape)

(544, 13)
(335, 13)
(209, 13)


In [232]:
#build a list of all points we want to checks
scr = screens[screens.lng.isnull()==False][['lat','lng']]

In [237]:
screen_list = list(zip(scr['lat'],scr['lng']))

In [263]:
find_closest((38.876059,-77.020240), screen_list)

8.65658792939013e-06

In [261]:
screens[screens['Screen Addresss (City)']=='Silver Spring']

Screen Design: Screen Design  \
193                  Discovery HQ Silver Spring Screen   
300                  Central Silver Spring - Grosvenor   
307                                      Silver Spring   
308  Engrain - The Fenwick - Silver Spring MD - 861...   

                Opportunity: Opportunity Name Screen Placement Address  \
193        Discovery HQ - Silver Spring PILOT        1 Discovery Place   
300         Central Silver Spring - Grosvenor           8455 Fenton St   
307                       MOCO 4 Regional SVC       One Veterans Place   
308  The Fenwick (8616 2nd Ave Silver Spring)             8616 2nd Ave   

    Screen Addresss (City) Screen Address (State) Screen Address (Zip)  \
193          Silver Spring                     MD                20910   
300          Silver Spring                     MD                20910   
307          Silver Spring                     MD                20910   
308          Silver Spring                     MD                20910   

     Install Stage Region    Screen Type        lat        lng  \
193  Customer Live     DC  Public Screen  38.995472 -77.028290   
300  Customer Live     MD  Public Screen  38.995071 -77.023972   
307  Customer Live     MD  Public Screen  38.997418 -77.024508   
308  Customer Live     MD  Public Screen  38.997318 -77.034363   

                                       Address close_date  
193   1 Discovery Place Silver Spring MD 20910        NaN  
300      8455 Fenton St Silver Spring MD 20910    9/25/17  
307  One Veterans Place Silver Spring MD 20910        NaN  
308        8616 2nd Ave Silver Spring MD 20910        NaN

In [302]:
find_within_distance((33.782524200000005, -84.380598800000001), screen_list, radius=1000000)

(33.782524200000005, -84.3805988)
found self (33.782524200000005, -84.380598800000001)


542

In [309]:
model_data['lat_lng'] = list(zip(model_data['lat'],model_data['lng']))

In [313]:
model_data['closest_ts']=np.nan

In [314]:
for i, row in model_data.iterrows():
    closest_ts = find_closest(row.lat_lng, screen_list)
    model_data.loc[model_data['id']==row.id, 'closest_ts'] = closest_ts

In [321]:
model_data.sort_values('closest_ts',ascending=False)

id                                               name  \
1004  0061a00000EV5MB              LinkedIn Singapore Office - Singapore   
1619  0061a00000G59Mn                            Clear Channel Singapore   
748   0061a00000HyoAr                                   Brookfield - TBD   
123   0061a00000MhvNV                Cushman & Wakefield-271 17th Street   
65    0061a00000MiXV4                                         ADP- Spain   
1531  0061a00000Fs3J7               Urban Investment Partners- Portfolio   
967   0061a00000MiJGe                               UIP - DC - Frequency   
961   0061a00000KkNd5                  STACK Real Estate- Salt Lake CIty   
731   0061a00000HwQJW     Elsey Partners - 117 27th Ave Apartments - MSP   
1020  0061a00000ItSRl                     Titan Development- Albuquerque   
1008  0061a00000Irjwn                                 Waarborg Vastgoed-   
1007  0061a00000Irjws                                 Waarborg Vastgoed-   
1621  0061a00000DUfbR                                          HDR DC HQ   
1174  0061a00000JqQ5u                      RREEF - DC - 1800 K Street NW   
985   0061a00000Iut45                  Jones Lang LaSalle -  Ireland-TBD   
1495  0061a00000ESulz                       CBRE - PHL - 1735 Market St.   
1021  0061a00000JqXOk            Decathlon-Amsterdam - Decathalon Office   
299   0061a00000GSJwb                    reThink - Central Florida- FDOT   
1623  0061a00000B0oEt                                  Accenture - Paris   
1421  0061a00000HFCy6                                      Kisio Digital   
904   0061a00000HGL3Y                Kilbourne Group- Woodrow Apartments   
672   0061a00000IrdzM                            Domain Companies - NOLA   
1636  0061a00000G4cch                                   JCDecaux - Paris   
1426  0061a00000Hy7oh      New Orleans Regional Transit Authority - NOLA   
129   0061a00000LSCz2                      Engie - Euroatrium Saint-Ouen   
1522  0061a00000IsJPG                      Ford Motor Company - Fordland   
1681  0061a00000AXzkG                   The Bernstein Companies Maryland   
1132  0061a00000KNrtv                 Wayne State University - DET - TBD   
1139  0061a00000KNhxY               Bedrock - Detroit - 630 Woodward Ave   
249   0061a00000G1cYB                         City of Detroit- City Hall   
...               ...                                                ...   
1250  0061a00000GFwW8                  BRP Development Corporation - NYC   
1516  0061a00000EGF6n                      Greystar - NJ - Parkway Lofts   
117   0061a00000KOUq5                               Bozzuto - Hubbard221   
265   0061a00000HyRoJ                    Binghamton- Binghamton U (SUNY)   
1612  0061a00000EIPrF                     Windsor - BOS - Cambridge Park   
344   0061a00000IskFY             Bozzuto - DC - The Modern at Art Place   
203   0061a00000HwLsg  George Washington University (GWU)- Funger Hal...   
1361  0061a00000EVh5T                  Forest City - The Lenox Club - VA   
187   0061a00000JqPiQ  Instrata Park Slope - Invesco Owned (Bozzuto M...   
1342  0061a00000G4hk0          Mill Creek- Modera Capitol Hill - SEATTLE   
1270  0061a00000ESp7K                          Schupp Comp - Hyatt Place   
201   0061a00000HwLwd  George Washington University- Mt. Vernon Campu...   
1262  0061a00000DJmlP                    Skanska - 101 Seaport Boulevard   
1261  0061a00000EVScz                Skanska USA - 121 Seaport Boulevard   
80    0061a00000IsDCs  Lincoln Property Company - Chicago - Niche 905...   
1225  0061a00000FsdKL                      Kettler - mflats Crystal City   
1183  0061a00000KN0sd                  Central Silver Spring - Grosvenor   
1160  0061a00000Js5Fy                                   Bozzuto- Central   
1121  0061a00000KMNAO                  Avalon North Point - AVB - Boston   
540   0061a00000KOTuB                     Mill Creek - NJ - Modera Lofts   
1080  0061a00000ItcM2                     Verve Mountain View - UDR

In [324]:
model_data['within_one_tenth']=np.nan
model_data['within_one_half']=np.nan
model_data['within_one']=np.nan
model_data['within_five']=np.nan

In [328]:
for i, row in model_data.iterrows():
    within_one_tenth = find_within_distance(row.lat_lng, screen_list, .1)
    within_one_half = find_within_distance(row.lat_lng, screen_list, .5)
    within_one = find_within_distance(row.lat_lng, screen_list, 1)
    within_five = find_within_distance(row.lat_lng, screen_list, 5)
    model_data.loc[model_data['id']==row.id, 'within_one_tenth'] = within_one_tenth
    model_data.loc[model_data['id']==row.id, 'within_one_half'] = within_one_half
    model_data.loc[model_data['id']==row.id, 'within_one'] = within_one
    model_data.loc[model_data['id']==row.id, 'within_five'] = within_five
    

In [331]:
model_data.head(100)

id                                               name  \
0   0061a00000GcQdL                          525 Water St SW- CMC Mgmt   
1   0061a00000Mi8dC                       Bozzuto - NJ - Park + Garden   
2   0061a00000M4PAU                              Greystar- The Cameron   
3   0061a00000MgelJ                             JLL- 2035 NW Front Ave   
4   0061a00000KOF8z     International Monetary Fund- Spring Conference   
5   0061a00000Mgpep                              GAP HQ- 1 Harrison St   
6   0061a00000HIaGb                         AVB - AVA Theater - Boston   
7   0061a00000Mgdqw                    Metro Village- Spring Place- DC   
8   0061a00000M4P5Q                  Greystar- Aventine at Fort Totten   
9   0061a00000M3j9z                 Wood Partners- Alta Roosevelt- CHI   
10  0061a00000M3DjE                Oxford Properties- Royal Bank Plaza   
11  0061a00000Kkodx  City of Cambridge- 119 Windsor St- Public Heal...   
12  0061a00000M4fhF                         The Collective- Bedford SQ   
13  0061a00000MiHRw                     Schupp Comp - Hyatt Place/AUSA   
14  0061a00000HyrMf                    reThink - Central Florida- FDOT   
15  0061a00000Irt5y                Daro Management - The Vintage on 16   
16  0061a00000M3b3g       Concord Hospitality- Hyatt House Jersey City   
17  0061a00000DJUWm               Seattle Children's Hospital: 12/2017   
18  0061a00000M6FZM               741 Harvard- Indigo Real Estate- SEA   
19  0061a00000M5oXh                     Carrolsburg: A Condominium- DC   
20  0061a00000KkG0w                          Modera Sandy Springs- ATL   
21  0061a00000KkGL1                     The Horizons- Gates Hudson- VA   
22  0061a00000KkY0j           Greystar- The Aspen- 1031 4th Street LLC   
23  0061a00000KkYER                  Donohoe- MD - 7101 Wisconsin Ave.   
24  0061a00000KkrTg      Transwestern - ATL - 64 Perimeter Center East   
25  0061a00000KNxoP                         Central Place Office lobby   
26  0061a00000HI1sq         DTZ-Polk and Taylor Property, LLC- renewal   
27  0061a00000Kjymq                     GE- 75 Fountain St, Providence   
28  0061a00000MiJPM                    City of Seattle- Justice Center   
29  0061a00000IuzTB           MRP - 3101 Wilson HART Hartford Bldg LLC   
..              ...                                                ...   
70  0061a00000IsTPK            City of Brookline - Brookline Town Hall   
71  0061a00000Iuzsa          CBRE / GE - San Ramon - GE Digital Office   
72  0061a00000Iuztd  CBRE/GE - Cincinnati - GE Global Operations Ce...   
73  0061a00000IsLBM                     GE - Norwalk - 801 Main Street   
74  0061a00000Hy4YO               Greystar - Solana Olde Town - Denver   
75  0061a00000GGPcs  Center for Global Development - DC - 2055 L St...   
76  0061a00000HvLpd                             Medstar Health - Pilot   
77  0061a00000ItSOX                           Camden- NoMa DC I and II   
78  0061a00000Ire2B                         GE- 41 Farnsworth - Boston   
79  0061a00000JqU7x                      GE- Chicago - 500 W Monroe St   
80  0061a00000IsDCs  Lincoln Property Company - Chicago - Niche 905...   
81  0061a00000HEsku   Equity Residential - Courthouse Plaza Apartments   
82  0061a00000GSDvb                                          Docker SF   
83  0061a00000GFx46                      GAP-Corporate Headquarters-CA   
84  0061a00000IuGuU               City of Flagstaff, AZ- City Hall TBD   
85  0061a00000Mi7W7    Outlook Development Group - Seven 10 West - CHI   
86  0061a00000OLXPu           The World Bank Group-HQ w/o hotels Pilot   
87  0061a00000OLQ9V                         JBC - 151 N Franklin - CHI   
88  0061a00000MhlY6                 Rose Associates - NYC - 88 Leonard   
89  0061a00000M4tLy                          Greystar- The Grand - DEN   
90  0061a00000JsKS0                 AREP - 700 E Pratt St / 111 Market   
91  0061a00000KkQV7  Related Companies - The Arlington - Boston (Ro...   
92  0061a00000M3OMt  

In [332]:
df.to_csv('model_data.csv', index=False)